text/0123-improve-span-status-api.md

```diff
@@ -0,0 +1,303 @@
+````
+
+
+
+## Internal details
+
+I expect a basic and reasonable implementation would be to define fields on the
+Span object, set them in the setter and then implement some getter for the
+exporters to use. A second possibility is to store the data into attributes or
+events. The choice of storage may have some modest effects on memory usage
+but primarily I expect the choice would be driven by the SDK API we want to
+read stored data back.
+
+It is also possible for the SDK to destructure the Exception data into simpler
+serializable types though I'd expect serialization is typically the domain of
+the exporter and there is a fair amount of policy involved in terms of what
+data is captured and how it is formatted for transport. There are definite
+risks that the end-to-end scenario will be less functional or less performant
+if SDKs intercede here.
+
+In some languages storing an exception with traceback could be very
+memory intensive. Python prevents locals in the callstack from being
+GC'ed. One suggestion is to use a synchronous callback to the exporter
+allowing it process the data immediately in some way that would lower
+the memory usage. For example the exporter could indicate the
+exception could be disposed after optionally extracting a fraction of the
+information to serialize.
+
+
+## Trade-offs and mitigations
+
+As mentioned in the motivation section, the issues of inconsistency, loss of
+fidelity and conversion difficulty are all on a sliding scale. I expect this
+design improves each of these issues at the expensive of some increased
+object-model/API complexity. Making the OpenTelemetry status representation
+more expressive also could cause inconsistency problems for the opposite reason
+- now there might be multiple reasonable representations for a status and the
+user becomes unclear which one to pick. In general I expect this to be mitigated
+with documented conventions, API defaults, and the potentially canonicalizing
+data anywhere in the processing pipeline. 
+
+One place I neglected to go further was defining additional types of error data
+format beyond string/int/Exception. This might encapsulate things such as
+key/value pairs (ala structured logging) or more complex or niche status types
+(COM IErrorInfo). There is nothing inherently problematic with them but I felt
+the increased expressiveness was getting diminishing returns. One mitigation to
+key/value pair data in particular would be to put auxilliary data in the span
+attributes using some convention, for example "Error.UserName"="Bob" might be
+used together with an error message string "Failed to find user {UserName}".
+Another mitigation might be adding language specific overloads that handle
+additional error types.
+
+The current proposal only gives one string which can be used either for a 
+freeform message or a textual status code. Adding a 2nd string to the
+StatusData would allow both to be collected SxS. This is another example of
```

Comment:
let's spell this abbrev. out to be more inclusive for non-racing sport enthusiasts and non-native speakers, sparing them the Google lookup I just had to do 😉
```suggestion
StatusData would allow both to be collected side by side. This is another example of
```

Reply:
Interesting, I never knew "SxS" was related to racing! I probably first saw the acronym at work but its been too long now for me to trace the origin. Happy to change it : )

Reply:
Well I didn't conduct any diligent research, the racing terms were just the first that popped up when I was googling for that term 😄